stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from ..core.experimental import experimental
+
+from tensorflow.keras.layers import Input, Dense, Lambda, Layer
+import tensorflow as tf
+from tensorflow.keras import backend as K
+
+__all__ = ["InfoMax"]
+
+
+class Discriminator(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][1], input_shapes[0][1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list or tuple of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+        features, summary = inputs
+
+        score = tf.linalg.matvec(features, tf.linalg.matvec(self.kernel, summary))
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+class InfoMax:
+    """
+    A class to wrap stellargraph models for Deep Graph Infomax unsupervised training
+    (https://arxiv.org/pdf/1809.10341.pdf).
+
+    Args:
+        base_model: the base stellargraph model class
+    """
+
+    _NODE_FEATS = "INFO_MAX_NODE_FEATURES"
+
+    def __init__(self, base_model):
+
+        if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+            raise TypeError(
+                f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+            )
+
+        self.base_model = base_model
+
+        # specific to full batch models
+        self.corruptible_inputs_idxs = [0]
+
+    def unsupervised_node_model(self):
```

Comment:
We have been trying to standardise this method and always call it `build()`. The latter is supposed to build the correct model given the information that it has, e.g., if base model is GCN then go for unsupervised full batch node model or if base model is GraphSAGE then go for unsupervised mini-batch node model, etc.